prov/verbs/src/verbs_cm.c

```diff
@@ -227,6 +272,90 @@ struct fi_ops_cm fi_ibv_msg_ep_cm_ops = {
 	.join = fi_no_join,
 };
 
+#ifdef INCLUDE_VERBS_XRC
+
+static int
+fi_ibv_msg_xrc_ep_connect(struct fid_ep *ep, const void *addr,
+		   const void *param, size_t paramlen)
+{
+	struct sockaddr *dst_addr;
+	void *adjusted_param;
+	struct fi_ibv_ep *_ep = container_of(ep, struct fi_ibv_ep,
+					     util_ep.ep_fid);
+	int ret;
+
+	if (!_ep->srqn) {
+		ret = ep->fid.ops->control(&ep->fid, FI_ENABLE, NULL);
+		if (ret)
+			return ret;
+	}
+	adjusted_param = fi_ibv_msg_alloc_xrc_params(param, &paramlen);
+	if (!adjusted_param)
+		return -errno;
+
+	dst_addr = rdma_get_peer_addr(_ep->id);
+	ret = fi_ibv_connect_xrc(_ep, dst_addr, 0, adjusted_param, paramlen);
+	free(adjusted_param);
+	return ret;
+}
+
+static int
+fi_ibv_msg_xrc_ep_accept(struct fid_ep *ep, const void *param, size_t paramlen)
+{
+	void *adjusted_param;
+	struct fi_ibv_ep *_ep =
+		container_of(ep, struct fi_ibv_ep, util_ep.ep_fid);
+	int ret;
+
+	if (!_ep->srqn) {
+		ret = ep->fid.ops->control(&ep->fid, FI_ENABLE, NULL);
+		if (ret)
+			return ret;
+	}
+
+	adjusted_param = fi_ibv_msg_alloc_xrc_params(param, &paramlen);
+	if (!adjusted_param)
+		return -errno;
+
+	ret = fi_ibv_accept_xrc(_ep, 0, adjusted_param, paramlen);
+	free(adjusted_param);
+	return ret;
+}
+
+#else /* INCLUDE_VERBS_XRC */
+
+static int
+fi_ibv_msg_xrc_ep_connect(struct fid_ep *ep, const void *addr,
+		   const void *param, size_t paramlen)
+{
+	/* Should not be callable if XRC is not enabled */
+	assert(0);
+	return -FI_ENOSYS;
+}
```

Comment:
If this function is called when INCLUDE_VERBS_XRC is not defined,  should we assert or should we return an error?
Do we want the provider to assert and take down the node?
Or do we want to just return an error?  Then it would be up to the application to abort or not use XRC.
You have it set to assert.
This logic is in several functions.

Reply:
When XRC is not enabled at libfabric configure (i.e. INCLUDE_VERBS_XRC is not defined because --with-verbs-xrc was not included), then it is a programming error on my part if these functions get called (as well as the other cases where functions are defined in the not INCLUDE_VERBS_XRC case). When XRC is not configured it should be impossible for the application to use the provider in a way that will take this code path. Beyond validating correctness, the real purpose these functions serve is to just limit the creep of XRC specific data/structures into the MSG endpoint code when XRC has not been configured. If/when XRC is always configured by default and optionally selectable by the application, then these functions would go away.